### PR TITLE
[PLU-102] feat(postman): add rich text editor for email body

### PR DESCRIPTION
**Pull request by stanleynguyen**
## Problem

Basic rich text editor without variable support

## Solution

Use tiptap. Kudos to @pregnantboy for previous work from postman editor (I copy homework only)

## Tests

- [x] Basic content with all formatting

Screenshot 2023-10-18 at 4 37 41 PM

Screenshot 2023-10-18 at 4 38 11 PM

- [x] required rule from hook form (will prevent the form from "Continue"-ing)
- [x] non-empty placeholder

### HTML Tags

tag | supported | currently used (in how many steps) | remarks
---|---|---|---
span | x | 5
b | x | 100
strong | x | 24
i | x | 25
em | x | 3
u | x | 30
ins | | 0
br | x | 40
p | x | 75
ul | x | 5
ol | x |1
li | x | 9
h1 | x | 4
h2 | x | 4
h3 | x | 1
h4 | x | 1
h5 | | 0
h6 | | 0
a | x | 39
img | x | 20
div | x | 2
tbody | x | 15
table | x | 28
tr | x | 29
td | x | 29
th | x | 7
sup | | 0
caption | | 1 | not supported, but tested and works fine
mark | | 0

## Deploy Notes

N/A

### CODE REVIEW

File: packages/frontend/src/components/RichTextEditor/index.tsx [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+      TableCell.configure({
+        HTMLAttributes: {
+          style: 'border:1px solid black;',
+        },
+      }),
+      Placeholder.configure({
+        placeholder,
+      }),
+    ],
+    content: initialValue,
```
Comment (ogp-weeloong) [inline packages/frontend/src/components/RichTextEditor/index.tsx:59]: It looks like prosemirror is stripping out unknown tags. E.g. create a pipe with something like :
```
<p>first para</p>
<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/2/2f/Google_2015_logo.svg/272px-Google_2015_logo.svg.png" />
<p>2nd para</p>
```
The image is stripped out, and data is lost if user tries to edit the step.

Is there a way we can configure PM to display unknown HTML as raw Text

**Review by ogp-weeloong — CHANGES_REQUESTED** [review]
Gotta figure out the missing `img` tag problem

File: packages/frontend/src/components/RichTextEditor/index.tsx [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  const { control } = useFormContext()
+  return (
+    <FormControl>
+      {label && (
+        <FormLabel isRequired={required} description={description}>
+          {label}
+        </FormLabel>
+      )}
+      <Controller
+        rules={{ required: true }}
```
Comment (ogp-weeloong) [inline packages/frontend/src/components/RichTextEditor/index.tsx]: why no follow the prop? :o

File: packages/frontend/src/components/RichTextEditor/index.tsx [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+const RichTextEditor = ({
+  required,
+  defaultValue,
+  name,
+  label,
+  description,
+  disabled,
+  placeholder,
+}: {
+  required?: boolean
```
Comment (ogp-weeloong) [inline packages/frontend/src/components/RichTextEditor/index.tsx]: nit: can we add a type for this e.g. `RichTextEditorProps`

File: packages/frontend/src/components/RichTextEditor/index.tsx [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+import { MenuBar } from './MenuBar'
+
+const Editor = ({
+  onChange,
+  initialValue,
+  editable,
+  placeholder,
+}: {
+  onChange: (...event: any[]) => void
```
Comment (ogp-weeloong) [inline packages/frontend/src/components/RichTextEditor/index.tsx]: nit: new `EditorProps` type

File: packages/frontend/src/components/RichTextEditor/MenuBar.tsx [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    onClick: (editor: Editor) => editor.chain().focus().undo().run(),
+  },
+  {
+    label: 'Redo',
+    icon: <RiArrowGoForwardLine />,
+    onClick: (editor: Editor) => editor.chain().focus().redo().run(),
+  },
+]
+
+export const MenuBar = ({ editor }: { editor: Editor | null }) => {
```
Comment (ogp-weeloong) [inline packages/frontend/src/components/RichTextEditor/MenuBar.tsx]: nit: `MenuBarProps` type pls?

File: packages/frontend/src/components/RichTextEditor/MenuBar.tsx [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+  {
+    label: 'Add row',
+    icon: <RiInsertRowBottom />,
+    onClick: (editor: Editor) => editor.chain().focus().addRowAfter().run(),
+  },
+  {
+    label: 'Remove row',
+    icon: <RiDeleteRow />,
+    onClick: (editor: Editor) => editor.chain().focus().deleteRow().run(),
+  },
```
Comment (ogp-weeloong) [inline packages/frontend/src/components/RichTextEditor/MenuBar.tsx:167]: nit

```suggestion
  },
  {
    label: 'Delete table',
    icon: <RiDeleteBinLine />,
    onClick: (editor: Editor) => editor.chain().focus().deleteTable().run(),
    isDisabled: (editor: Editor) => !editor.isActive('table'),
  },
```

Wondering if we can also have a delete table button?

File: packages/frontend/src/components/RichTextEditor/MenuBar.scss [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    fill: currentColor;
+    height: 100%;
+    width: 100%;
+  }
+
+  &:hover,
+  &.is-active {
+    background-color: #303030;
+  }
+}
```
Comment (ogp-weeloong) [inline packages/frontend/src/components/RichTextEditor/MenuBar.scss:34]: Nit

```suggestion
    &.is-disabled {
      color: #aa004b;
      cursor: default;
    }
}
```

Small feature request to add `isDisabled`, ignore if you think too heavy

Comment (ogp-weeloong) [reply packages/frontend/src/components/RichTextEditor/MenuBar.scss:34]: Actually nvm, I looked at this for a bit. A bit tricky, not worth.

e.g. `deleteColumn` is a no-op if there is only 1 column, so we want to disable that. But there doesnt' seem to be an easy way to get current column count.

File: packages/frontend/src/components/RichTextEditor/index.tsx (hunk 1 above)
Comment (stanleynguyen) [reply packages/frontend/src/components/RichTextEditor/index.tsx:59]: I think bopian. I thought our users dont need images but just did a look up in our db they actually need image support. Will be adding it

File: packages/frontend/src/components/RichTextEditor/MenuBar.tsx (hunk 6 above)
Comment (stanleynguyen) [reply packages/frontend/src/components/RichTextEditor/MenuBar.tsx:167]: Actually donnid cos they can just highlight the table and delete, vs for row and col they can't do that

File: packages/frontend/package.json [hunk 8]
```diff
@@ -34,6 +44,7 @@
     "react-intl": "6.2.7",
     "react-json-tree": "^0.18.0",
     "react-router-dom": "6.11.2",
+    "sass": "1.69.4",
```
Comment (ogp-weeloong) [inline packages/frontend/package.json:47]: forgot to mention - this should be dev dependency i think.

**Comment by ogp-weeloong** [comment]
Legit work on the comprehensive check against postman allowed tags!!!!

But paiseh poking again... I found visual regression though using this input (mostly edge cases involving nesting and invalid HTML such as block items in inline item (e.g. I put a `h1` in an `li`), or users' own custom styling). Can use html prettier to make it human readable.

```
<p>Hi thar I haz inline image between here <img src="https://file.go.gov.sg/spflogo.png" width="667"> and here </p><p>Other text...<p>nested para<p>nested para 2</p></p></p><table><caption>Test table</caption><tr><th>th1</th><th>th2</th><th>th3</th></tr><tbody><tr><td style="border:1px solid red;"><div style="color:red">topkek styled div <p><i>italics p</i> in div</p></div></td><td style="border:1px solid black;" colspan="2"><p>x</p></td></tr><tr><td><h1>Heading 1</h1><h2>heading 2</h2><p><h3>H3 in p </h3><i>italics thing again</i></p></td><td><p>qwe</p></td><td style="border:1px solid black;"><ol>ol no items </lo></td></tr><tr><td style="border:1px solid black;" colspan="1" rowspan="1"><p>zzz</p></td><td style="border:1px solid purple;"><ul><li>item 1</li><li>item 2</li><li><h1>heading 1 in item</h1></li></ul></td><td style="border:1px solid yellow;"><div>a div element!</div></td></tr></tbody></table>
```

#### Before
Screenshot 2023-10-25 at 10 16 37 PM

#### After
Screenshot 2023-10-25 at 10 19 00 PM

I feel like we really should make this opt-in. I think I have (off top of my head) way to make opt-in backwards compatible without too much complexity, wanna discuss tmr?

**Review by ogp-weeloong — APPROVED** [review]
LGTM! Discussed about tradeoff of preserving backward compatibility vs killing PowerInput with possibility of visual regressions.

Conclusion was that according to homework, regressions are very unlikely. We should kill PowerInput ASAP to reduce tech debt in the system.

**Pull request merged**